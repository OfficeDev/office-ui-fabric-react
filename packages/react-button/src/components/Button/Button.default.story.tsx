import * as React from 'react';
import { Button } from './Button';

export default {
  title: 'Fluent UI Core/Button',
  component: Button,
};

export const Default = (props) => <Button {...props} />;
Default.args = {
  children: 'Default',
};
