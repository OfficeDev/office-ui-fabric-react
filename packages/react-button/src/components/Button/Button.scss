/**
 * TODO: Once theming is updated, consider removing all of this.
 *
 * Note: All :root rules must be literals, OR references to variables
 * defined as literals in the same scope. If you refer to variables,
 * they will be resolved on the :root element only; not at usage time.
**/
:root {
  /* sizing */
  --button-size-smallest: 24px;
  --button-size-smaller: 24px;
  --button-size-small: 24px;
  --button-size-regular: 32px;
  --button-size-large: 40px;
  --button-size-larger: 48px;
  --button-size-largest: 64px;

  --button-padding: 0 24px;
  --button-height: var(--button-size-regular);
  --button-minWidth: 96px;
  --button-maxWidth: 280px;

  --button-contentGap: 10px;
  --button-iconSize: 16px;

  /* borders and shadow */
  --button-borderRadius: 2px;
  --button-borderWidth: 1px;
  --button-boxShadow: 0px 2px 4px -0.75px rgba(0, 0, 0, 0.1);
  --button-boxShadow-disabled: none;

  /* transforms */
  --button-transform-pressed: scale(0.95);
  --button-transition: transform 0.1s linear;

  /* TODO: Remove. Fonts should only be defined in the theme. */
  --button-fontFamily: 'Segoe UI', 'Helvetica Neue', 'Apple Color Emoji', 'Segoe UI Emoji', Helvetica, Arial, sans-serif;
  --button-fontSize: 14px;
  --button-fontWeight: 600;

  /* TODO: Remove. Colors should only be defined in the theme. */
  --button-background: white;
  --button-background-hovered: rgb(237, 235, 233);
  --button-background-focused: var(--button-background);
  --button-background-pressed: rgb(225, 223, 221);
  --button-background-disabled: rgb(237, 235, 233);
  --button-contentColor: rgb(44, 38, 33);
  --button-contentColor-hovered: var(--button-contentColor);
  --button-contentColor-focused: var(--button-contentColor);
  --button-contentColor-pressed: var(--button-contentColor);
  --button-contentColor-disabled: rgb(200, 198, 196);
  --button-borderColor: rgb(225, 223, 221);
  --button-borderColor-disabled: transparent;
  --button-icon: inherit;
  --button-iconColor-hovered: var(--button-icon);
  --button-iconColor-pressed: var(--button-icon);
  --button-iconColor-focused: var(--button-icon);
  --button-iconColor-disabled: var(--button-contentColor-disabled);

  --button-focusColor: #000;
  --button-focusInnerColor: #fff;

  /* TODO: Remove. Accent should only be defined in the theme. */
  --accent-fill: rgb(98, 100, 167);

  --accent-background-focused: var(--accent-fill);
  --accent-background-hovered: rgb(88, 90, 150);
  --accent-background-pressed: rgb(70, 71, 117);
  --accent-background-disabled: rgb(237, 235, 233);

  --accent-contentColor: white;
  --accent-contentColor-hovered: var(--accent-contentColor);
  --accent-contentColor-focused: var(--accent-contentColor);
  --accent-contentColor-pressed: var(--accent-contentColor);
  --accent-contentColor-disabled: rgb(200, 198, 196);
  --accent-borderColor: transparent;

  --accent-icon: inherit;
  --accent-iconColor-hovered: var(--accent-icon);
  --accent-iconColor-pressed: var(--accent-icon);
  --accent-iconColor-focused: var(--accent-icon);
  --accent-iconColor-disabled: var(--accent-contentColor-disabled);
}

/**
  * Slot definitions
  *
  * Slot definitions must align to the names of the slots themselves. They should:
  *
  * 1. Contain the actual css property definitions (E.g. background-color: {value})
  * 2. Property definitions should map to tokens using the format: {component}-{property}
  * 3. Avoid hard coding values unless there are absolutely no use cases to change
  *    the property value.
  * 4. Order hardcoded css values first, tokens based second. This way we can identify
  *    what is hardcoded easily.
 **/

.root {
  align-items: center;
  border-style: solid;
  display: inline-flex;
  justify-content: center;
  outline: none;
  position: relative;
  user-select: none;
  box-sizing: border-box;

  background: var(--button-background, white);
  color: var(--button-contentColor, rgb(44, 38, 33));

  border-color: var(--button-borderColor);
  border-radius: var(--button-borderRadius);
  border-width: var(--button-borderWidth);
  box-shadow: var(--button-boxShadow);
  height: var(--button-height);
  padding: var(--button-padding);
  transition: var(--button-transition);
  white-space: var(--button-whiteSpace, nowrap);

  /**
  * Modifiers
  *
  * Modifiers must align to the props on the component. They should:
  *
  * 1. Redefine the values for {component}-{property} tokens.
  * 2. Map property definitions to variables using the format: {component}-{property}
  * 3. Avoid defining styles directly.
 **/
  &.primary {
    --button-fill: var(--accent-fill);
    --button-fill-hovered: var(--accent-fill-hovered);
    --button-fill-pressed: var(--accent-fill-pressed);

    --button-text: var(--accent-text);
    --button-text-hovered: var(--accent-text-hovered);
    --button-text-pressed: var(--accent-text-pressed);

    --button-borderColor: var(--accent-borderColor);

    --button-icon: var(--accent-icon);
    --button-icon-hovered: var(--accent-icon-hovered);
    --button-icon-pressed: var(--accent-icon-pressed);
  }

  &.iconOnly {
    width: var(--button-height);
    --button-padding: 0;
  }

  &.circular {
    --button-borderRadius: 50000px;
    --button-minWidth: var(--button-height);
  }

  &.fluid {
    width: 100%;
    max-width: 100%;
  }

  &.size_smallest {
    --button-height: var(--button-size-smallest);
  }

  &.size_smaller {
    --button-height: var(--button-size-smaller);
  }

  &.size_small {
    --button-height: var(--button-size-small);
  }

  &.size_large {
    --button-height: var(--button-size-large);
  }

  &.size_larger {
    --button-height: var(--button-size-larger);
  }
  &.size_largest {
    --button-height: var(--button-size-largest);
  }

  &.disabled {
    pointer-events: none;
    background-color: var(--button-fill-disabled);
    color: var(--button-text-disabled);
    border-color: var(--button-borderColor-disabled);
    box-shadow: var(--button-boxShadow-disabled);

    &.icon {
      color: var(--button-icon-disabled, var(--button-text-disabled));
    }
  }
}

.content {
  font-family: var(--button-fontFamily);
  font-size: var(--button-fontSize);
  font-weight: var(--button-fontWeight);
}

.icon {
  color: var(--button-icon);
  display: flex;
  align-items: center;
  justify-content: center;
  width: var(--button-iconSize);
  height: var(--button-iconSize);
  font-size: var(--button-iconSize, inherit);
}

.loader {
  /* TODO: styling once we have a sample. */
}

/* selectors */

.root > *:not(:first-child) {
  margin-left: var(--button-contentGap);
}

.root:hover {
  background: var(--button-background-hovered, var(--button-background));
  color: var(--button-contentColor-hovered, var(--button-contentColor));
  border-color: var(--button-borderColor-hovered, var(--button-borderColor));
  box-shadow: var(--button-boxShadow-hovered, var(--button-boxShadow));

  .icon {
    color: var(--button-iconColor-hovered, var(--button-contentColor-hovered));
  }
}

.root:active {
  background: var(--button-background-pressed, var(--button-background-hovered));
  color: var(--button-contentColor-pressed, var(--button-contentColor-hovered, var(--button-contentColor)));
  border-color: var(--button-borderColor-pressed, var(--button-borderColor-hovered, var(--button-borderColor)));
  box-shadow: var(--button-boxShadow-pressed, var(--button-boxShadow-hovered, var(--button-boxShadow)));

  transform: var(--button-transform-pressed);
  transition: var(--button-transition-pressed);

  .icon {
    color: var(--button-iconColor-pressed, var(--button-contentColor-pressed));
  }
}

.root:focus::after {
  position: absolute;
  left: -1px;
  top: -1px;
  right: -1px;
  bottom: -1px;
  content: '';
  box-sizing: border-box;
  border-radius: var(--button-borderRadius);
  border: var(--button-focus-width, 2px) solid var(--button-focusColor);
  box-shadow: 0 0 0 var(--button-focus-innerWidth, 1px) var(--button-focusInnerColor) inset;
}
