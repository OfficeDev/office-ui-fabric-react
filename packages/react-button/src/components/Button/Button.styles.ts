import { makeStyles } from '@fluentui/react-theme-provider';

export const useButtonStyles = makeStyles({
  displayName: 'Button',
  root: [
    'ms-Button',
    {
      cursor: 'pointer',
      alignItems: 'center',
      borderStyle: 'solid',
      display: 'inline-flex',
      justifyContent: 'center',
      outline: 'none',
      position: 'relative',
      userSelect: 'none',
      boxSizing: 'border-box',
      verticalAlign: 'middle',

      background: 'var(--button-background, white)',
      color: 'var(--button-contentColor)',

      borderColor: 'var(--button-borderColor)',
      borderRadius: 'var(--button-borderRadius)',
      borderWidth: 'var(--button-borderWidth)',
      boxShadow: 'var(--button-boxShadow)',

      width: 'var(--button-width)',
      height: 'var(--button-height)',
      minHeight: 'var(--button-minHeight)',

      paddingLeft: 16, // 'var(--button-padding)',
      paddingRight: 16, // 'var(--button-padding)',
      paddingTop: 0, // 'var(--button-padding)',
      paddingBottom: 0, // 'var(--button-padding)',

      transition: 'var(--button-transition)',
      whiteSpace: 'var(--button-whiteSpace, nowrap)',

      fontFamily: 'var(--button-fontFamily)',
      fontSize: 'var(--button-fontSize)',

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      fontWeight: 'var(--button-fontWeight)' as any,

      // @include focus-border(
      //   $padding: '-1px',
      //   $color: 'var(--button-focusColor)',
      //   $thickness: 'var(--button-focusWidth, 2px)',
      //   $radius: 'var(--button-borderRadius',
      // );
      // @includeFocusBoxShadowHOffset_0VOffset_0Blur_0Thickness_1pxColor: 'var(--button-focusInnerColor))',

      ['& > *:not(:first-child)']: {
        marginLeft: 'var(--button-contentGap)',
      },

      '&:hover': {
        background: 'var(--button-hovered-background, var(--button-background))',
        color: 'var(--button-hovered-contentColor, var(--button-contentColor))',
        borderColor: 'var(--button-hovered-borderColor, var(--button-borderColor))',
        boxShadow: 'var(--button-hovered-boxShadow, var(--button-boxShadow))',

        '.ms-Button-icon': {
          color: 'var(--button-hovered-iconColor, var(--button-iconColor))',
        },
      },
    },
  ],

  icon: [
    'ms-Button-icon',
    {
      color: 'var(--button-iconColor)',
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      width: 'var(--button-iconSize)',
      height: 'var(--button-iconSize)',
      fontSize: 'var(--button-iconSize, inherit)',
      fontWeight: 'normal',
    },
  ],

  _primary: {
    '--button-background': 'var(--accent-background)',
    '--button-borderColor': 'var(--accent-borderColor)',
    '--button-contentColor': 'var(--accent-contentColor)',
    '--button-icon-color': 'var(--accent-iconColor)',

    '--button-hovered-background': 'var(--accent-hovered-background)',
    '--button-hovered-borderColor': 'var(--accent-hovered-borderColor)',
    '--button-hovered-contentColor': 'var(--accent-hovered-contentColor)',
    '--button-hovered-iconColor': 'var(--accent-hovered-iconColor)',

    '--button-focused-background': 'var(--accent-focused-background)',
    '--button-focused-borderColor': 'var(--accent-focused-borderColor)',
    '--button-focused-contentColor': 'var(--accent-focused-contentColor)',
    '--button-focused-iconColor': 'var(--accent-focused-iconColor)',

    '--button-pressed-background': 'var(--accent-pressed-background)',
    '--button-pressed-borderColor': 'var(--accent-pressed-borderColor)',
    '--button-pressed-contentColor': 'var(--accent-pressed-contentColor)',
    '--button-pressed-iconColor': 'var(--accent-pressed-iconColor)',

    '--button-disabled-background': 'var(--accent-disabled-background)',
    '--button-disabled-border-color': 'var(--accent-disabled-borderColor)',
    '--button-disabled-content-color': 'var(--accent-disabled-contentColor)',
    '--button-disabled-icon-color': 'var(--accent-disabled-iconColor)',
  },

  _iconOnly: {
    '--button-width': 'var(--button-height, var(--button-minHeight))',
    '--button-padding': 0,
  },

  _circular: {
    '--button-border-radius': 50000,
  },

  _fluid: {
    width: '100%',
    maxWidth: '100%',
  },

  // eslint-disable-next-line @typescript-eslint/naming-convention
  _size_smallest: {
    '--button-min-height': 'var(--button-size-smallest)',
  },

  // eslint-disable-next-line @typescript-eslint/naming-convention
  _size_smaller: {
    '--button-min-height': 'var(--button-size-smaller)',
  },

  // eslint-disable-next-line @typescript-eslint/naming-convention
  _size_small: {
    '--button-min-height': 'var(--button-size-small)',
  },

  // eslint-disable-next-line @typescript-eslint/naming-convention
  _size_large: {
    '--button-min-height': 'var(--button-size-large)',
  },

  // eslint-disable-next-line @typescript-eslint/naming-convention
  _size_larger: {
    '--button-min-height': 'var(--button-size-larger)',
  },

  // eslint-disable-next-line @typescript-eslint/naming-convention
  _size_largest: {
    '--button-min-height': 'var(--button-size-largest)',
  },

  _disabled: {
    pointerEvents: 'none',
    opacity: 'var(--button-disabled-opacity)',
    backgroundColor: 'var(--button-disabled-background)',
    contentColor: 'var(--button-disabled-contentColor)',
    borderColor: 'var(--button-disabled-borderColor)',
    boxShadow: 'var(--button-disabled-boxShadow)',

    '.ms-Button-icon': {
      color: 'var(--button-disabled-iconColor)',
    },
  },
});
