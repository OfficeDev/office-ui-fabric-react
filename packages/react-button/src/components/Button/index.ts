export * from './Button.types';
export * from './Button.styles';
export * from './useButton';
export * from './useButtonState';
export * from './Button';
