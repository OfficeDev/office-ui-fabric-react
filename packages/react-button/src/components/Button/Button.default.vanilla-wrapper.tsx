import * as React from 'react';
import { Button } from './Button';

export default {
  title: 'Fluent UI Core/Button',
  component: Button,
};

export const Vanilla = (p) => <Button {...p}>Vanilla</Button>;
