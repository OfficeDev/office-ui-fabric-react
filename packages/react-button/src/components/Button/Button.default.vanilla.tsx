import * as React from 'react';
import { Button } from './Button';

export default (p) => <Button {...p}>Vanilla</Button>;
