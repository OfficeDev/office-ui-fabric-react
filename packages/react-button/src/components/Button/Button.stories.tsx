import * as React from 'react';
import { Button } from './Button';
import { ButtonProps } from './Button.types';
import * as classes from './Button.stories.scss';

const ButtonVariants = (props: ButtonProps) => (
  <div className={classes.hStack}>
    <Button {...props} content="Hello, world" icon="X" />
    <Button {...props} primary content="Hello, world" icon="X" />
    <Button {...props} disabled content="Hello, world" icon="X" />
    <Button {...props} primary disabled content="Hello, world" icon="X" />
  </div>
);

export const ButtonCss = () => (
  <>
    <h3>A button can appear round using the `circular` prop.</h3>
    <ButtonVariants circular />

    <h3>A button can fill the width of its container using the `fluid` prop.</h3>
    <div className={classes.vStack}>
      <Button fluid content="Hello, world" icon="X" />
      <Button fluid primary content="Hello, world" icon="X" />
      <Button fluid disabled content="Hello, world" icon="X" />
      <Button fluid primary disabled content="Hello, world" icon="X" />
    </div>

    <h3>A button can contain only an icon using the `iconOnly` prop.</h3>
    <ButtonVariants iconOnly />

    <h3>A button can be both `circular` and `iconOnly`.</h3>
    <ButtonVariants circular iconOnly />

    <h3>An icon button can format its Icon to appear before or after its content.</h3>
    <div className={classes.vStack}>
      <ButtonVariants iconPosition="before" />
      <ButtonVariants iconPosition="after" />
    </div>

    {/* <h3>A button can inherit its background and have a subtle appearance using the `inverted` prop.</h3>
    <ButtonVariants inverted /> */}

    <h3>A button can show a loading indicator using the `loading` prop.</h3>
    <ButtonVariants loading />

    <h3>A button can be sized.</h3>
    <div className={classes.vStack}>
      <ButtonVariants size="smallest" />
      <ButtonVariants size="smaller" />
      <ButtonVariants size="small" />
      <ButtonVariants size="large" />
      <ButtonVariants size="larger" />
      <ButtonVariants size="largest" />
    </div>
  </>
);

export const ButtonTokens = () => (
  <>
    <h3>A button can be colored using inline tokens.</h3>
    <div className={classes.vStack}>
      <Button
        icon="O"
        tokens={{
          padding: '3px 10px',
          fontSize: '12px',
          fontWeight: '600',
          borderColor: {
            default: 'rgba(27, 31, 35, 0.2)',
            hovered: 'rgba(27, 31, 35, .5)',
            pressed: 'rgba(27, 31, 35, .5)',
          },
          contentGap: '4px',
          borderRadius: '4px',
          borderWidth: '1px',
          background: {
            default: 'linear-gradient(-180deg,#34d058,#28a745 90%)',
            hovered: 'linear-gradient(-180deg,#34d058,#28a745 90%)',
            pressed: '#279f43',
          },
          contentColor: {
            default: 'white',
            hovered: 'white',
            pressed: 'white',
          },
          transform: {
            pressed: 'none',
          },
        }}
        content="Github: Open issue"
      />
      <Button
        tokens={{
          fontFamily: `"Amazon Ember", Arial, sans-serif`,
          fontSize: '13px',
          fontWeight: '400',
          borderRadius: '3px',
          background: {
            default: 'linear-gradient(to bottom,#f7dfa5,#f0c14b)',
            hovered: 'linear-gradient(to bottom,#f5d78e,#eeb933)',
            pressed: '#f0c14b',
          },
          contentColor: {
            default: 'rgb(17, 17, 17)',
          },
          borderColor: {
            default: 'rgb(168, 135, 52) rgb(156, 126, 49) rgb(132, 106, 41)',
            hovered: '#a88734 #9c7e31 #846a29',
          },
          padding: '0 10px',
        }}
        content="Amazon: Proceed to checkout"
      />

      <Button
        tokens={{
          background: {
            default: 'rgb(229, 9, 20)',
            hovered: 'rgb(255, 50, 61)',
            pressed: 'rgb(229, 9, 20)',
          },
          contentColor: {
            default: 'rgb(255, 255, 255)',
            hovered: 'rgb(255, 255, 255)',
            pressed: 'rgb(255, 255, 255)',
          },
          borderColor: {
            default: 'rgb(255, 255, 255)',
          },
          borderWidth: '0px',
          padding: '7px 17px',
          borderRadius: '3px',
          fontFamily: '"Netflix Sans", "Helvetica Neue", Helvetica, Arial, sans-serif',
          fontSize: '16px',
          fontWeight: '400',
        }}
        content="Netflix: Sign In"
      />
      <Button
        tokens={{
          height: '48px',
          transform: {
            pressed: 'none',
          },
          background: {
            default: 'rgb(29, 185, 84) none repeat scroll 0% 0% / auto padding-box border-box',
            hovered: 'rgb(30, 215, 96) none repeat scroll 0% 0% / auto padding-box border-box',
            pressed: '#1aa34a',
          },
          contentColor: {
            default: 'rgb(255, 255, 255)',
            hovered: 'rgb(255, 255, 255)',
            pressed: 'rgb(255, 255, 255)',
          },
          borderColor: {
            default: 'rgb(255, 255, 255)',
            hovered: 'rgb(255, 255, 255)',
          },
          borderWidth: '0px',
          padding: '17px 48px',
          borderRadius: '500px',
          fontFamily: 'Circular, Helvetica, Arial, sans-serif',
          fontSize: '14px',
          fontWeight: '700',
        }}
        content="Spotify: GET PREMIUM"
      />
      <Button
        tokens={{
          height: '52px',
          background: {
            default: 'rgba(0, 0, 0, 0) none repeat scroll 0% 0% / auto padding-box border-box',
          },
          contentColor: {
            default: 'rgb(0, 0, 0)',
          },
          borderColor: {
            default: 'rgb(0, 0, 0)',
          },
          borderWidth: '2px',
          padding: '17px 48px',
          borderRadius: '500px',
          fontFamily: 'Circular, Helvetica, Arial, sans-serif',
          fontSize: '14px',
          fontWeight: '700',
        }}
        content="Spotify: LEARN MORE"
      />
    </div>
    <h3>A tokenized button can be customized for any size or padding.</h3>
    <div className={classes.vStack}>
      <ButtonVariants
        tokens={{
          height: '24px',
          fontSize: '12px',
          iconSize: '22px',
          padding: '0 8px',
        }}
      />
      <ButtonVariants
        tokens={{
          height: '70px',
          fontSize: '48px',
          iconSize: '48px',
        }}
      />
    </div>
  </>
);
