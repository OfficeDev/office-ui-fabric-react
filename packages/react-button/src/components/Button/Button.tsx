import * as React from 'react';
import { useInlineTokens } from '@fluentui/react-theme-provider';
import { renderButton } from './renderButton';
import { useButton } from './useButton';
import {
  useButtonStyles,
  // useStaticButtonStyles,
  useButtonContentStyles,
  // useStaticButtonContentStyles,
  useButtonIconStyles,
  // useStaticButtonIconStyles,
  ButtonClassNames,
} from './useButtonClasses';
import { ButtonProps } from './Button.types';

/**
 * Define a styled Button, using the `useButton` hook.
 * {@docCategory Button}
 */
export const Button = React.forwardRef<HTMLElement, ButtonProps>((props, ref) => {
  const state = useButton(props, ref);

  state.className = useButtonStyles(
    state,
    {
      componentNames: 'Button',
      tokens: state.tokens,
    },
    ButtonClassNames.root,
    state.className,
  );
  /* eslint-disable @typescript-eslint/no-explicit-any */
  (state.content as any).className = useButtonContentStyles(
    state,
    {
      tokens: state.tokens,
    },
    ButtonClassNames.content,
    (state.content as any).className,
  );
  (state.icon as any).className = useButtonIconStyles(
    state,
    {
      tokens: state.tokens,
    },
    ButtonClassNames.icon,
    (state.icon as any).className,
  );
  /* eslint-enable @typescript-eslint/no-explicit-any */

  useInlineTokens(state, '--button');

  return renderButton(state);
});

Button.displayName = 'Button';
