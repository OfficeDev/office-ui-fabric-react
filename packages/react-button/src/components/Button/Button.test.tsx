import * as React from 'react';
import { Button as ButtonComponent } from './Button';
import * as renderer from 'react-test-renderer';
import { mount, ReactWrapper } from 'enzyme';
import { isConformant } from '../../common/isConformant';
import { MergeStylesProvider } from '@fluentui/react-theme-provider';
import { ButtonProps } from './Button.types';

/** Use merge-styles provider to ensure styles show up in snapshots. */
const Button = React.forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => (
  <MergeStylesProvider>
    <ButtonComponent {...props} ref={ref} />
  </MergeStylesProvider>
));

describe('Button', () => {
  let wrapper: ReactWrapper | undefined;

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
  });

  isConformant({
    Component: ButtonComponent,
    displayName: 'Button',
  });

  /**
   * Note: see more visual regression tests for Button in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const component = renderer.create(<Button>Default button</Button>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders anchor when href prop is provided', () => {
    const component = renderer.create(<Button href="https://www.bing.com">Default button</Button>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('can be focused', () => {
    const rootRef = React.createRef<HTMLButtonElement>();

    wrapper = mount(<Button ref={rootRef}>Focus me</Button>);

    expect(typeof rootRef.current).toEqual('object');
    expect(document.activeElement).not.toEqual(rootRef.current);

    rootRef.current?.focus();

    expect(document.activeElement).toEqual(rootRef.current);
  });
});
