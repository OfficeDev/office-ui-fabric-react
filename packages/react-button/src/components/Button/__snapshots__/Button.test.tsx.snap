// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders a default state 1`] = `
<button
  className=" ms-Button a1k6fduh a122n59 atuwxu6 a4d9j23 a10pi13n a1xqy1su a1ewtqcl amrv4ls a1lzbuw2 aim6dvu a1f4c47i a1iedgaj aw1n0wr a70odt1 a1vhwvfz a1cz3ezt a1oddkp7 a13kh4da act4cwl ahp4f9l aeo0lyr a1whg1k0 a7uf1io azb318i a1hju0nx a1k1mxt7 a1egkakw aofteie a16j5cmg a1745qu5 a1q4sedi a1uyijub a1o7jddm a1648qya a1qr8kil adhe3jx a1ow1jxr a15t8rk3 a1qruv96 a33d3hg a19jhgz2 ajd3y3t a24elfs a1uo23zi a13lmprv a123g821 a196fwmc aydxa82 aldjma3 agsy9ym a1gk5ei5 a1qoqgcf a57year a11leut1 a195mrzw adg5kxv a8v4fns a1s6fcnf afwnqy8 avpzubt a16c20uz avjb07u azkkow9 acdblym ag706s2 ajik90z a1cz6fjg a1c5wrov akjzyke a1tdbhw7 a1w5j6e3 a1ehx0w9 asy0rpt a14qiakg a1ik6wn0 a8tzcgd a1velgez a135zgpe a18swz6d axvxnzq a1g55moa axdv31r a1nh3o6k agacytd abwah6s auv3803 arfyqzs a1bi4j9d a1hwvjsp anqe8ok a1qcrciv a1w5wezo aeujwd0 a3615mr anmlcw7 ailih6g aqcp0tr a7i9xs0 a1mdaal6 a1i4tqxa a2hmzya a4va83 a15jls3h a1aehjj5 a18cyma4"
>
  <span
    className=" ms-Button-content ap6vxd"
  >
    Default button
  </span>
</button>
`;

exports[`Button renders anchor when href prop is provided 1`] = `
<a
  className=" ms-Button a1k6fduh a122n59 atuwxu6 a4d9j23 a10pi13n a1xqy1su a1ewtqcl amrv4ls a1lzbuw2 aim6dvu a1f4c47i a1iedgaj aw1n0wr a70odt1 a1vhwvfz a1cz3ezt a1oddkp7 a13kh4da act4cwl ahp4f9l aeo0lyr a1whg1k0 a7uf1io azb318i a1hju0nx a1k1mxt7 a1egkakw aofteie a16j5cmg a1745qu5 a1q4sedi a1uyijub a1o7jddm a1648qya a1qr8kil adhe3jx a1ow1jxr a15t8rk3 a1qruv96 a33d3hg a19jhgz2 ajd3y3t a24elfs a1uo23zi a13lmprv a123g821 a196fwmc aydxa82 aldjma3 agsy9ym a1gk5ei5 a1qoqgcf a57year a11leut1 a195mrzw adg5kxv a8v4fns a1s6fcnf afwnqy8 avpzubt a16c20uz avjb07u azkkow9 acdblym ag706s2 ajik90z a1cz6fjg a1c5wrov akjzyke a1tdbhw7 a1w5j6e3 a1ehx0w9 asy0rpt a14qiakg a1ik6wn0 a8tzcgd a1velgez a135zgpe a18swz6d axvxnzq a1g55moa axdv31r a1nh3o6k agacytd abwah6s auv3803 arfyqzs a1bi4j9d a1hwvjsp anqe8ok a1qcrciv a1w5wezo aeujwd0 a3615mr anmlcw7 ailih6g aqcp0tr a7i9xs0 a1mdaal6 a1i4tqxa a2hmzya a4va83 a15jls3h a1aehjj5 a18cyma4"
  href="https://www.bing.com"
  role="button"
>
  <span
    className=" ms-Button-content ap6vxd"
  >
    Default button
  </span>
</a>
`;
